File: 백준/Gold/16234. 인구 이동/인구 이동.cc
#include <bits/stdc++.h>
using namespace std;

int n, l, r, ans, sum, a[50][50], visited[50][50], dy[] = { 0, 0, 1, -1 }, dx[] = { 1, -1, 0, 0 };
vector<pair<int, int>> v;

void dfs(int y, int x) {
	visited[y][x] = 1;
	v.push_back({ y, x });
	sum += a[y][x];

	for (int i = 0; i < 4; i++) {
		int ny = y + dy[i];
		int nx = x + dx[i];
		if (ny >= 0 && ny < n && nx >= 0 && nx < n && !visited[ny][nx] && abs(a[ny][nx] - a[y][x]) >= l && abs(a[ny][nx] - a[y][x]) <= r) 
			dfs(ny, nx);
	}
}
 
int main()
{
	ios_base::sync_with_stdio(false);
	cin.tie(NULL);
	cout.tie(NULL);

	cin >> n >> l >> r;

	for (int i = 0; i < n; i++) {
		for (int j = 0; j < n; j++)
			cin >> a[i][j];
	}

	while (true) {
		bool flag = false;
		memset(visited, 0, sizeof(visited));
		for (int i = 0; i < n; i++) {
			for (int j = 0; j < n; j++) {
				if (!visited[i][j]) {
					v.clear();
					sum = 0;
					dfs(i, j);
					for (auto k : v) {
						a[k.first][k.second] = sum / v.size();
					}
					if (v.size() >= 2)
						flag = true;
				}
			}
		}

		if (!flag) {
			cout << ans;
			break;
		}

		ans++;
	}



	return 0;
}

File: 백준/Gold/16234. 인구 이동/인구 이동.java
import java.io.*;
import java.util.*;

public class Main {
	static int n, l, r, ans;
	static int[][] a;
	static boolean[][] visited;
	static List<Node> list;
	static int[] dy = {-1, 0, 1, 0};
	static int[] dx = {0, 1, 0, -1};

	public static void main(String[] args) throws IOException {
		BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
		BufferedWriter bw = new BufferedWriter(new OutputStreamWriter(System.out));

		StringTokenizer st = new StringTokenizer(br.readLine());
		n = Integer.parseInt(st.nextToken());
		l = Integer.parseInt(st.nextToken());
		r = Integer.parseInt(st.nextToken());

		a = new int[n][n];
		for (int i = 0; i < n; i++) {
			st = new StringTokenizer(br.readLine());
			for (int j = 0; j < n; j++) {
				a[i][j] = Integer.parseInt(st.nextToken());
			}
		}

		while (true) {
			boolean flag = false;
			visited = new boolean[n][n];
			for (int i = 0; i < n; i++) {
				for (int j = 0; j < n; j++) {
					if (!visited[i][j]) {
						list = new ArrayList<>();
						int sum = dfs(i, j);
						if (list.size() != 1) {
							flag = true;
							int tmp = sum / list.size();
							for (Node node : list) {
								a[node.y][node.x] = tmp;
							}
						}
					}

				}
			}

			if (!flag)
				break;

			ans++;
		}

		bw.write(ans + "");

		bw.flush();
	}

	static int dfs(int y, int x) {
		visited[y][x] = true;
		int ret = a[y][x];
		list.add(new Node(y, x));
		for (int i = 0; i < 4; i++) {
			int ny = y + dy[i];
			int nx = x + dx[i];
			if (ny < 0 || nx < 0 || ny >= n || nx >= n)
				continue;
			if (visited[ny][nx] || Math.abs(a[y][x] - a[ny][nx]) < l || Math.abs(a[y][x] - a[ny][nx]) > r)
				continue;
			ret += dfs(ny, nx);
		}

		return ret;
	}

	static class Node {
		int y, x;

		Node(int y, int x) {
			this.y = y;
			this.x = x;
		}
	}
}
